logstash-core/lib/logstash/plugin_metadata.rb

```diff
@@ -0,0 +1,108 @@
+# encoding: utf-8
+
+require 'thread_safe/cache'
+
+module LogStash
+  ##
+  # `PluginMetadata` provides a space to store key/value metadata about a plugin, typically metadata about
+  # external resources that can be gleaned during plugin registration.
+  #
+  # Data is persisted across pipeline reloads, and no effort is made to clean up metadata from pipelines
+  # that no longer exist after a pipeline reload.
+  #
+  #  - It MUST NOT be used to store processing state
+  #  - It SHOULD NOT be updated frequently.
+  #  - Individual metadata keys MUST be Symbols and SHOULD NOT be dynamically generated
```

Comment:
I don't know how important it is that the interface for PluginMetadata remain the same between Java and Ruby plugins, but Java plugins can't use symbols as keys.